src/goto-instrument/synthesizer/expr_enumerator.h

```diff
@@ -0,0 +1,511 @@
+    return expected_size;
+  }
+
+protected:
+  /// Size of expected enumerated expressions. The size of a leaf expression
+  /// is 1. So we enumerate only when the `expected_size` is 1.
+  unsigned int expected_size;
+  exprst leaf_exprs;
+};
+
+/// Non-leaf enumerator enumerates expressions of form
+/// f(a_1, a_2, ..., a_n)
+/// where a_i's are leaf expressions enumerated by sub-enumerators.
+class non_leaf_enumeratort : public enumerator_baset
+{
+public:
+  non_leaf_enumeratort() : enumerator_baset()
+  {
+  }
+
+  /// \param enumerators a list of enumerator (e_1,...,e_n),
+  /// where a_i in the expressions enumerated by this enumerator can be
+  /// enumerated by the enumerator e_i.
+  /// \param is_none if it is a none enumerator.
+  explicit non_leaf_enumeratort(
+    const enumeratorst &enumerators,
+    bool is_none = false)
+    : enumerator_baset(is_none)
+  {
+    /// initialize the list of sub-enumerators
+    sub_enumerators = enumeratorst();
+    for(const auto &g : enumerators)
+    {
+      sub_enumerators.push_back(g->clone());
+    }
+
+    expected_size = 0;
+
+    /// Arity n is the number of sub-enumerators
+    arity = sub_enumerators.size();
+    INVARIANT(arity > 0, "non-zero arity");
+
+    /// Only enumerate sub-expressions for good size partitions.
+    /// This is an optimization to avoid enumerating equivalent expressions.
+    /// As default: Use no optimization. Accept all partitions.
+    is_good_partition = [](const partitiont &) { return true; };
+  }
+
+  /// Given a list `enumerators` of enumerators, return the caresian product
+  /// of expressions enumerated by each enumerator in the list.
+  std::list<exprst> cartesian_product_of_enumerators(
+    const enumeratorst &enumerators,
+    unsigned int pos = 0);
+
+  /// Enumerate all partitions of n into k components.
+  /// Order of partitions is considered relevant.
+  std::list<partitiont> get_partitions(unsigned int n, unsigned int k);
+
+  exprst enumerate() override
+  {
+    exprst result = exprst();
+
+    // Enumerate nothing when `expected_size` is too small to be partitioned.
+    if(expected_size - 1 < arity)
+      return result;
+
+    // For every possible partition, set `expected_size` of
+    // each sub-enumerator to be the corresponding component in the partition.
+    for(const auto &partition : get_partitions(expected_size - 1, arity))
+    {
+      if(!is_good_partition(partition))
+        continue;
+
+      set_sub_enumerator_sizes(partition);
+
+      // Compute the cartesian product as result.
+      for(const auto &product_tuple :
+          cartesian_product_of_enumerators(sub_enumerators))
+      {
+        // Optimization: rule out equivalent expressions
+        // using certain equivalence class.
+        // Keep only representation tuple of each equivalence class.
+        if(is_equivalence_class_representation(product_tuple))
+          result.push_back(instantiate(product_tuple));
+      }
+    }
+
+    return result;
+  }
```

Comment:
Most functions (not just this one) should be moved to the implementation file. This just pollutes the header file, making the API very hard to read.